contrib/tools/workflowcheck/README.md

```diff
@@ -0,0 +1,252 @@
+by the workflow. It is highly optimized to scan large codebases quickly.
+
+**NOTE: This will not catch all cases of non-determinism such as global var mutation. This is just a helper and
+developers should still scrutinize workflow code for other non-determinisms.**
+
+## Installing
+
+Build like a normal Go app. With this repository cloned and [Go](https://golang.org/) installed and on the `PATH`, run:
+
+To install with [Go](https://golang.org/) installed and on the `PATH`, run:
+
+    go install go.temporal.io/sdk/contrib/tools/workflowcheck
+
+Or you can simply build by running `go build` inside of this directory. 
+
+## CLI Usage
+
+The executable has arguments in the form:
+
+    workflowcheck [-flag] [package]
+
+To simply check all source, navigate to the root of the repository/module to check and, with `workflowcheck` on
+the `PATH`, run:
+
+    workflowcheck ./...
+
+For each workflow passed to `RegisterWorkflow` that is considered non-deterministic, this will output the reasons why.
+There is no output if all workflows are properly deterministic.
+
+The output is hierarchical if the reason occurs in code that the workflow calls. For example, if a workflow X calls Y
+which calls Z that iterates over a map, the result will show X calling Y calling Z and why Z itself was
+non-deterministic. This helps for debugging and disabling false positives.
+
+Note, the source must be valid and compilable Go source for the tool to work. The package arguments are the same as any
+other `go` tool. Run `workflowcheck -help` for more details on arguments. See "Configuration" below on how to configure
+the checker.
+
+### Example
+
+Given the following file called `workflow.go` at the root of the `example.com/myworkflow` module:
+
+```go
+package myworkflow
+
+import (
+	"fmt"
+	"time"
+
+	"go.temporal.io/sdk/worker"
+	"go.temporal.io/sdk/workflow"
+)
+
+func Register(w worker.WorkflowRegistry) {
+	w.RegisterWorkflow(MyWorkflow)
+}
+
+func MyWorkflow(workflow.Context) error {
+	now := time.Now()
+	fmt.Printf("Now: %v\n", now)
+	return nil
+}
+```
+
+If `workflowcheck .` is run, the following output will appear:
+
+```
+/path/to/workflow.go:12:2: example.com/myworkflow.MyWorkflow is non-deterministic, reason: calls non-deterministic function time.Now
+  time.Now is non-deterministic, reason: declared non-deterministic
+/path/to/workflow.go:12:2: example.com/myworkflow.MyWorkflow is non-deterministic, reason: calls non-deterministic function fmt.Printf
+  fmt.Printf is non-deterministic, reason: accesses non-deterministic var os.Stdout
+```
+
+The `-show-pos` argument can be provided to see exactly where each offending call happens. If we wanted to allow
+`fmt.Printf` to be used in all workflows, we could have the config `workflowcheck.config.yaml`:
+
+```yaml
+decls:
+  fmt.Printf: false
+```
+
+Then running `workflowcheck -config workflowcheck.config.yaml .` Would not only show the `time.Now` error. See the
```

Comment:
I'd have the tool look for config in a default location

Reply:
Good question/thought!

I considered that, but not only is configuration not required, you are not going to run this tool centrally, it is very dependent on context. I had considered something like a `.workflowcheck.config.yaml` but often for analysis tools that is not for where the program is executed but rather where the program is scanning. But there is more complexity and little value in merging many configs in different packages based on where this is scanning (we have inline ignoring for most of those use cases).

Since I could not figure out whether the implicit config was based on scan location or run location, I decided it was harmless to ask for a config explicitly if needed. Most people probably won't need a config (our entire `samples-go` repo will not for example).